### PR TITLE
Deprecate `_Result.excinfo`

### LINKED ISSUE #86: trigger a deprecationwarning for Result.excinfo
https://github.com/pytest-dev/pluggy/blob/1037dc9615fda1bece07169fbd3c61956ef01b26/pluggy/callers.py#L32-L34

we kept the code for pytest, but we should at least deprecate it and remove the use in future

### PR DESCRIPTION
**Pull request by tgoodlet**
Resolves #86

### CODE REVIEW

File: pluggy/callers.py [hunk 1]
```diff
@@ -31,12 +31,17 @@ def __init__(self, result, excinfo):
 
     @property
     def excinfo(self):
+        """Get the exception info for this hook call (DEPRECATED in favor of
+        ``get_result()``).
+        """
+        msg = 'Use `get_result()` which raises the underlying exception'
```
Comment (RonnyPfannschmidt) [inline pluggy/callers.py]: i believe for clarity of the warning the property should be referenced in the text somehow

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt @nicoddemus pushed the change hopefully it's good enough :+1:

**Comment by tgoodlet** [comment]
@nicoddemus what's the recommended way to make CI happy here?

I kind of want to keep failure on unexpected deprecation warnings but we also can't wrap `pytest` runs with `pytest.deprecated_call()`?

**Comment by RonnyPfannschmidt** [comment]
@tgoodlet i beleive we need to set up pytest awrning filters in a way that we dont error on plugy warnings and make a pytest release where we dont fail anymore

**Comment by nicoddemus** [comment]
I think @RonnyPfannschmidt's suggestion is on point. 👍

**Comment by tgoodlet** [comment]
@nicoddemus gonna rebase this and put in a `newsfragment` using the new `towncrier` stuff :+1:

**Comment by tgoodlet** [comment]
@nicoddemus @RonnyPfannschmidt also do you both mind checking I did the `warnings` filter right?
It seems to work in CI at least.

**Comment by nicoddemus** [comment]
Awesome!

**Comment by nicoddemus** [comment]
Strange, I would expect the warnings to show only once, not for every call...

File: changelog/155.removal.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Deprecate ``_Result.excinfo`` in favor of ``_Result.get_result()``.
```
Comment (nicoddemus) [inline changelog/155.removal.rst]: Missing a `D` here 😉

**Comment by nicoddemus** [comment]
Just checking the problematic code in pytest.

```python
@hookimpl(hookwrapper=True)
def pytest_pyfunc_call(pyfuncitem):
    check_xfail_no_run(pyfuncitem)
    outcome = yield
    passed = outcome.excinfo is None
    if passed:
        check_strict_xfail(pyfuncitem)
```

becomes:

```python
@hookimpl(hookwrapper=True)
def pytest_pyfunc_call(pyfuncitem):
    check_xfail_no_run(pyfuncitem)
    outcome = yield
    try:
        outcome.get_result()
        passed = True
    except Exception:
        passed = False
    if passed:
        check_strict_xfail(pyfuncitem)
```

TBH doesn't look like an improvement to me. Or am I missing something?

(To be clear, totally in favor about the deprecation of `.result`)

File: changelog/155.removal.rst (hunk 2 above)
Comment (tgoodlet) [reply changelog/155.removal.rst]: Wait what?

**Comment by tgoodlet** [comment]
@nicoddemus I also think it's a little uglier but @RonnyPfannschmidt seemed convinced in #86.

**Comment by tgoodlet** [comment]
@nicoddemus btw we can also chat now on the new [gitter channel](https://gitter.im/pytest-dev/pytest)!

**Comment by tgoodlet** [comment]
@RonnyPfannschmidt still waiting on your input for this one.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus the question does become if we should expose the actual exception, from my pov, the concrete values of exception/result should be hidden away,

the following questions come up to me
1. does a single use really warrant exposing the data as is?
2. are there other uses that may fit an api that exposes an abstract success/failure? concept?

**Comment by nicoddemus** [comment]
> does a single use really warrant exposing the data as is?

You mean a single use in pytest, or in a more general manner, like a single use case (verify if any exception has occurred)?

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus in that case single use of the value in pytest - i believe its important to consider point 2 in relation

**Comment by nicoddemus** [comment]
A [quick search](https://github.com/search?p=4&q=outcome.excinfo&type=Code) shows that the only use case is indeed if a plugin is interested only if an exception has occurred or not.

TBH if we were discussing about including the API for `excinfo` I would agree we should look at a few more use cases first, but given we are talking about deprecating and then removing, I would like to see the use cases where keeping it is harmful (which is clearly the case of `.result`). It is just that so far, the few uses where checking if an exception has occurred will have to be rewritten, without a clear gain AFAICT.

**Comment by RonnyPfannschmidt** [comment]
i find that reasoning acceptable, on principle i'd still prefer if we had a success/failure indicator instead of exposing the exception

**Comment by nicoddemus** [comment]
That would work for pytest, but I did see some code in that search I posted before that were checking for a specific type of exception.

Also, we need to realize that we can always deprecate `.excinfo` later if we find a compelling use case to remove it.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus do you have a link, at least on github most of the results are accidentally committed tox virtualenvs

**Comment by nicoddemus** [comment]
Yeah those generate a lot of noise. I could find a few instances mostly in pytest plugins, for example: 

https://github.com/vodik/pytest-exceptional/blob/c8a9901618ea860b10967db741280c64a9f3b4a7/pytest_exceptional.py#L68

https://github.com/jankatins/pytest-error-for-skips/blob/f1c7daafed26930fd2f59f3d1baf0df5fdb59ead/pytest_error_for_skips.py#L11

https://github.com/cielavenir/pytest-forcefail/blob/81ad226cae191fe0b28798739ca8dbbf09ff81d4/pytest_forcefail.py#L19

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus all examples you linked use `get_result` and use pytest callinfo

**Comment by nicoddemus** [comment]
Oh you are right. Anyway, the code in those links will still need to be updated eventually... well I will leave up to you then, I don't have any strong arguments anymore. 👍

**Comment by RonnyPfannschmidt** [comment]
https://github.com/Authentise/pytest-raises/blob/aaa4999d3d79f3a07dfd6f8b6a6c01df827fecd2/pytest_raises/pytest_raises.py#L24
https://github.com/vsevolod-fedorov/hyperapp/blob/6e54afad8876114b2e5ad793c92ec576075ad2b8/conftest.py#L33

https://github.com/search?q=%22outcome.excinfo+is+None%22&type=Code

as per those its in actual usage (and github is just really bad to search code)

**Comment by RonnyPfannschmidt** [comment]
however it should be noted that all uses of the code would tremendously benefit from a failure indication

**Comment by RonnyPfannschmidt** [comment]
based on the information we collected i believe its only fair to keep the api,
but eventually replace it with a python3 api since exceptions can carry tracebacks there

we should also consider introducing a failure indication as an orthogonal matter

**Comment by nicoddemus** [comment]
> we should also consider introducing a failure indication as an orthogonal matter

You have a suggestion? I think two read-only properties, `passed` and `failed` might do the trick.

**Comment by nicoddemus** [comment]
Sounds good! 👍

**Comment by tgoodlet** [comment]
Closing according to the consensus that we keep `.excinfo` until a better *failure indication* API is proposed to replace it.

**Pull request closed without merging**